File: valid_sequence_from_root/is_valid_sequence.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.right = right


def is_valid_sequence(root: TreeNode, arr: List[int]) -> bool:
    """
    Time: O(n)
    Space: O(1)
    """
    return dfs(root, 0, arr)


def dfs(node: TreeNode, index: int, arr: List[int]):
    if not node or index >= len(arr) or arr[index] != node.val:
        return False
    if not node.left and not node.right:
        return len(arr) == index + 1
    return (dfs(node.left, index + 1, arr) or
            dfs(node.right, index + 1, arr))

File: valid_sequence_from_root/is_valid_sequence.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Time: O(n)
 * Space: O(1) 
 */
func isValidSequence(root *TreeNode, arr []int) bool {
	return dfs(root, 0, arr)
}

func dfs(node *TreeNode, depth int, arr []int) bool {
	if node == nil || depth >= len(arr) || node.Val != arr[depth] {
		return false
	}
	if node.Right == nil && node.Left == nil {
		return depth+1 == len(arr)
	}
	return dfs(node.Left, depth+1, arr) || dfs(node.Right, depth+1, arr)
}

func main() {

}
